Doc/c-api/unicode.rst

```diff
@@ -1502,3 +1502,77 @@ They all return ``NULL`` or ``-1`` if an exception occurs.
    :c:func:`PyUnicode_InternInPlace`, returning either a new Unicode string
    object that has been interned, or a new ("owned") reference to an earlier
    interned string object with the same value.
+
+PyUnicodeWriter
+^^^^^^^^^^^^^^^
+
+The :c:type:`PyUnicodeWriter` API can be used to create a Python :class:`str`
+object.
+
+.. versionadded:: 3.14
+
+.. c:type:: PyUnicodeWriter
+
+   An Unicode writer instance.
+
+   The instance must be destroyed by :c:func:`PyUnicodeWriter_Finish` on
+   success, or :c:func:`PyUnicodeWriter_Discard` on error.
+
+.. c:function:: PyUnicodeWriter* PyUnicodeWriter_Create(Py_ssize_t length)
+
+   Create an Unicode writer instance.
+
+   Set an exception and return ``NULL`` on error.
+
+.. c:function:: PyObject* PyUnicodeWriter_Finish(PyUnicodeWriter *writer)
+
+   Return the final Python :class:`str` object and destroy the writer instance.
+
+   Set an exception and return ``NULL`` on error.
+
+.. c:function:: void PyUnicodeWriter_Discard(PyUnicodeWriter *writer)
+
+   Discard the internal Unicode buffer and destroy the writer instance.
+
+.. c:function:: int PyUnicodeWriter_WriteChar(PyUnicodeWriter *writer, Py_UCS4 ch)
+
+   Write the single Unicode character *ch* into *writer*.
+
+   Return ``0`` on success, or set an exception and return ``-1`` on error.
+
+.. c:function:: int PyUnicodeWriter_WriteUTF8(PyUnicodeWriter *writer, const char *str, Py_ssize_t size)
+
+   Decode the string *str* from UTF-8 in strict mode and write the output into *writer*.
```

Comment:
I would prefer to have an error parameter for this as well, so that it's possible to use the standard codec error handling. This makes it easier to deal with data from sources which may have slightly broken data.

Reply:
Also: It should be documented what happens in case the UTF-8 decoding fails, i.e. does the writer unwind the whole addition, just up to the point where it failed decoding, etc. and where the writer position ends up after such an error.

Reply:
This PR implements the API decided and approved by the C API Working Group.

Use `PyUnicode_DecodeUTF8()` + `PyUnicodeWriter_WriteStr()` for other error handlers.

Reply:
Why does it need to be undefined ? It's not unusual run into errors when decoding raw C strings from other sources and building Unicode strings dynamically. In fact, it may not even be possible to restart the operation (think Unix pipes or reading from network buffers).

Reply:
It does not need to be undefined. It's just that currently, the behavior is undefined. Maybe some characters have been written, maybe not. It's just more efficient to leave it undefined, the writer doesn't have to be atomic and remind where it was.

Reply:
> In fact, it may not even be possible to restart the operation (think Unix pipes or reading from network buffers).

You can say that about any operation, if Python raises an exception, you can lose data.

Reply:
I modified the implementation to make WriteUTF8() atomic. Either it writes the full string, or it writes nothing.